Capabilities 2025 chapter (#4355)
HTTPArchive/almanac.httparchive.org: src/content/en/2025/capabilities.md

On mobile, usage jumped from 2.3% to 12.3%, and on desktop from 2.7% to 14.0%. This steep rise aligns with the API becoming [widely supported across all major engines](https://web.dev/blog/compressionstreams) in the last two years, removing a technical blocker and letting developers drop JavaScript polyfills and rely on native gzip/deflate compression.

The API is particularly appealing for data-heavy applications where streaming efficiency matters, which explains its strong adoption curve.

### Clipboard API

The [Clipboard API](https://developer.mozilla.org/docs/Web/API/Clipboard_API) provides asynchronous read and write access to the system clipboard. It supports plain text, HTML, images, and other formats, though support may vary across browsers.

[Security restrictions](https://developer.mozilla.org/docs/Web/API/Clipboard_API#security_considerations) require clipboard operations to be triggered by a user gesture (like a click).

```js
// Write text
await navigator.clipboard.writeText("Hello from Web Almanac!");

// Read text
const text = await navigator.clipboard.readText();

console.log(text); // "Hello from Web Almanac!"
```

> Supported in Chromium-based browsers and Safari. Firefox has partial support.

Reviewer: +1, I also stumbled upon this. A big cross-browser support gap is [Web-custom formats](https://developer.chrome.com/blog/web-custom-formats-for-the-async-clipboard-api).

> Supported in Chromium-based browsers and Safari. Firefox has partial support.

Reviewer 2: I added this clarification: "The Async Clibpoard API is supported in Chromium-based browsers, Safari and Firefox. Only Chromium-based browsers have support for richer clipboard data, like web custom formats."

Since indeed, I think I overlook Safari support for the `ClipboardItem`. Thanks for spotting it!

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: Capabilities
description: Capabilities chapter of the 2025 Web Almanac covering brand-new, powerful web platform APIs that give web apps access to hardware interfaces, enhance web-based productivity apps, built-in AI and more.
hero_alt: Hero image of Web Almanac characters with superhero capes plugging various capabilities into a web page.
authors: [Dawntraoz, MichaelSolati]

> reviewers: [webmaxru]

Reviewer: ```suggestion
reviewers: [webmaxru, tomayac]
```

In the age of AI, browsers cannot afford to be left behind—they must propose sustainable, accessible solutions for all, in order to democratize the use of AI. Consequently, we will see the initial use of these new Chrome-specific APIs in the Capabilities chapter this year.

## Methodology

This chapter, as in previous years, used the HTTP Archive's public dataset of millions of pages. These pages were archived as both desktop and mobile, since some sites serve different content based on what device is requesting the page.

### How does the HTTP Archive detect capabilities?

The HTTP Archive crawler parses the source code for all of these pages to determine which APIs were (potentially) used on the pages using regular expressions, such as `/navigator\.share\s*\(/g`.

> The way it works can cause some problems when it comes to detecting things: it may underreport some APIs used as it can't detect code that may exist due to minification, for example, when navigator was minified to `n;` or it may overreport occurrences of APIs because it doesn't run code to see if an API is actually used.

Reviewer: ```suggestion
The way it works can cause some problems when it comes to detecting things: it may underreport some APIs used as it can't detect code that may exist due to minification, for example, when navigator was minified to `n`; or it may overreport occurrences of APIs because it doesn't run code to see if an API is actually used.
```

### Compression Streams API

The [Compression Streams API](https://developer.mozilla.org/docs/Web/API/Compression_Streams_API) allows web apps to compress and decompress data using widely supported formats like GZIP and Deflate, directly in the browser. This enables more efficient transfer and storage of large data without relying on server-side processing.

Data is processed via `CompressionStream` and `DecompressionStream` objects, which integrate with the web's streaming APIs (`ReadableStream`, `WritableStream`).

```js
const text = "Hello Web Almanac 2025!";
const stream = new Blob([text]).stream();
const compressed = stream.pipeThrough(new CompressionStream("gzip"));
const decompressed = compressed.pipeThrough(new DecompressionStream("gzip"));
const result = await new Response(decompressed).text();

console.log(result); // "Hello Web Almanac 2025!"
```

> Since May 2023, this feature works across the latest devices and browser versions. Available in Chromium-based browsers, Safari and Firefox, but might not work in older devices or browsers.

Reviewer: ```suggestion
Since May 2023, this feature works across the latest devices and browser versions. It's available in Chromium-based browsers, Safari and Firefox, but might not work on older devices or other browsers.
```

On mobile, usage jumped from 2.3% to 12.3%, and on desktop from 2.7% to 14.0%. This steep rise aligns with the API becoming [widely supported across all major engines](https://web.dev/blog/compressionstreams) in the last two years, removing a technical blocker and letting developers drop JavaScript polyfills and rely on native gzip/deflate compression.

The API is particularly appealing for data-heavy applications where streaming efficiency matters, which explains its strong adoption curve.

### Clipboard API

The [Clipboard API](https://developer.mozilla.org/docs/Web/API/Clipboard_API) provides asynchronous read and write access to the system clipboard. It supports plain text, HTML, images, and other formats, though support may vary across browsers.

[Security restrictions](https://developer.mozilla.org/docs/Web/API/Clipboard_API#security_considerations) require clipboard operations to be triggered by a user gesture (like a click).

```js
// Write text
await navigator.clipboard.writeText("Hello from Web Almanac!");

// Read text
const text = await navigator.clipboard.readText();

console.log(text); // "Hello from Web Almanac!"
```

> Supported in Chromium-based browsers and Safari. Firefox has partial support.

Reviewer: ```suggestion
The Async Clibpoard API is supported in Chromium-based browsers and Safari. Firefox has partial support.
```

<figure>
  <table>
    <thead>
      <tr>
        <th>Built-in AI API</th>
        <th>Desktop</th>
        <th>Mobile</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>LanguageDetector</td>
        <td>0.28%</td>
        <td>0.26%</td>
      </tr>
      <tr>
        <td>Prompt</td>
        <td>0.08%</td>
        <td>0.09%</td>
      </tr>
      <tr>
        <td>Translator</td>
        <td>0.28%</td>
        <td>0.26%</td>
      </tr>
      <tr>
        <td>Summarizer</td>
        <td>0.13%</td>
        <td>0.14%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Built-in AI API usage", sheets_gid="843125108", sql_file="fugu.sql") }}</figcaption>
</figure>

Despite their availability, usage across the web remains very limited. As shown in the table below, each of these APIs appears on well under 1% of pages in both desktop and mobile datasets. LanguageDetector and Translator are the most commonly observed, each used by roughly 0.28% of desktop pages and 0.26% of mobile pages, while Prompt and Summarizer show even smaller footprints.

> The low adoption rates are expected. These APIs are new, often still evolving, and currently supported by a limited set of browsers, Chromium-only. Their inclusion in the 2025 dataset is nonetheless significant: it marks the first measurable presence of browser-native AI primitives in the HTTP Archive, establishing a baseline for tracking how built-in AI capabilities evolve on the web in future years.

Reviewer: ```suggestion
The low adoption rates are expected. These APIs are new, often still evolving, and currently supported by a limited set of browsers, Chrome and Edge. Their inclusion in the 2025 dataset is nonetheless significant: it marks the first measurable presence of browser-native AI primitives in the HTTP Archive, establishing a baseline for tracking how built-in AI capabilities evolve on the web in future years.
```
